src/main/java/io/cdap/plugin/jms/common/JMSConfig.java

```diff
@@ -0,0 +1,250 @@
+  // Schema
+  public static final String MESSAGE_ID = "messageId";
+  public static final String MESSAGE_TIMESTAMP = "messageTimestamp";
+  public static final String CORRELATION_ID = "correlationId";
+  public static final String REPLY_TO = "replyTo";
+  public static final String DESTINATION = "destination";
+  public static final String DELIVERY_MODE = "deliveryNode";
+  public static final String REDELIVERED = "redelivered";
+  public static final String TYPE = "type";
+  public static final String EXPIRATION = "expiration";
+  public static final String PRIORITY = "priority";
+
+  // Params
+  public static final String NAME_CONNECTION_FACTORY = "connectionFactory";
+  public static final String NAME_JMS_USERNAME = "jmsUsername";
+  public static final String NAME_JMS_PASSWORD = "jmsPassword";
+  public static final String NAME_PROVIDER_URL = "providerUrl";
+  public static final String NAME_TYPE = "type";
+  public static final String NAME_DESTINATION = "destination";
+  public static final String NAME_JNDI_CONTEXT_FACTORY = "jndiContextFactory";
+  public static final String NAME_JNDI_USERNAME = "jndiUsername";
+  public static final String NAME_JNDI_PASSWORD = "jndiPassword";
+  public static final String NAME_MESSAGE_TYPE = "messageType";
+  public static final String NAME_SCHEMA = "schema";
+
+  @Name(NAME_CONNECTION_FACTORY)
+  @Description("Name of the connection factory.")
+  @Nullable
+  @Macro
+  private String connectionFactory;
+
+  @Name(NAME_JMS_USERNAME)
+  @Description("Username to connect to JMS.")
+  @Macro
+  private String jmsUsername;
+
+  @Name(NAME_JMS_PASSWORD)
+  @Description("Password to connect to JMS.")
+  @Macro
+  private String jmsPassword;
+
+  @Name(NAME_PROVIDER_URL)
+  @Description("The URL of the JMS provider. For example, in case of an ActiveMQ Provider, the URL has the format " +
+    "tcp://hostname:61616.")
+  @Macro
+  private String providerUrl;
+
+  @Name(NAME_TYPE)
+  @Description("Queue or Topic.")
+  private String type;
+
+  @Name(NAME_DESTINATION)
+  @Description("Destination Queue/Topic name.")
+  @Macro
+  private String destination;
+
+  @Name(NAME_JNDI_CONTEXT_FACTORY)
+  @Description("Name of the JNDI context factory. For example, in case of an ActiveMQ Provider, the JNDI Context " +
+    "Factory is: org.apache.activemq.jndi.ActiveMQInitialContextFactory.")
+  @Nullable
+  @Macro
+  private String jndiContextFactory;
+
+  @Name(NAME_JNDI_USERNAME)
+  @Description("User name for the JNDI.")
+  @Nullable
+  @Macro
+  private String jndiUsername;
+
+  @Name(NAME_JNDI_PASSWORD)
+  @Description("Password for the JNDI.")
+  @Nullable
+  @Macro
+  private String jndiPassword;
+
+  @Name(NAME_MESSAGE_TYPE)
+  @Description("Supports the following message types: Message, Text, Bytes, Map.")
+  private String messageType;
+
+  @Name(NAME_SCHEMA)
+  @Nullable
+  @Description("Specifies the schema of the records outputted from this plugin.")
+  private String schema;
```

Comment:
why cant schema be macro?

Reply:
At schema, only the `payload` field is expected to take different data types. `payload` data type is automatically changed corresponding to the `messageType` the user selects. Since the `messageType` is already a macro, I didn't specify macro at schema.

Should we allow the users to manually customise the output schema (add new fields, manually change data types,...). This isn't specified in the PRD.

Reply:
could you please followup with @bdmogal on the product requirements? I think this should be allowed to be a macro.